src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,171 @@
+        Node<K, V> newNode = null;
+        int index = index(key);
+        Objects.checkIndex(index, tables.length);
+        if (key == null) {
+            key = (K) Integer.valueOf(index);
+        }
 
+        if (tables[index] == null) {
+            newNode = new Node<>(index, key, value, null);
+            tables[index(key)] = newNode;
+            if (size++ > threshold) {
+                resize();
+            }
+            return;
+        }
+        newNode = new Node<>(index, key, value, null);
+        if (newNode.getHash() == tables[index].hash) {
+            Node<K, V> tmpTables = tables[index];
+            while (tmpTables != null) {
+                if (newNode.key.equals(tmpTables.key)) {
+                    tmpTables.setValue(newNode.getValue());
+                    return;
+                }
+                tmpTables = tmpTables.next;
+            }
+        }
+        newNode.next = tables[index].next;
+        tables[index].next = newNode;
+        if (size++ > threshold) {
+            resize();
+        }
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            key = (K) Integer.valueOf(0);
+        }
+        int index = index(key);
+        if (index < 0) {
+            index = index * (-1);
+        }
+        Node<K, V> tmpNode;
+        for (tmpNode = tables[index]; tmpNode != null; tmpNode = tmpNode.next) {
+            if (tmpNode.getKey().equals(key)) {
+                return tmpNode.getValue();
+            }
+
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private Node<K, V>[] resize() {
+        Node<K, V>[] oldTable = tables;
+        Node<K, V>[] newTable = new Node[oldTable.length * 2];
+
+        int index;
+        for (int i = 0; i < oldTable.length; ++i) {
+            Node<K, V> tmp;
+            if ((tmp = oldTable[i]) != null) {
+                oldTable[i] = null;
+                if (tmp.next == null) {
+                    index = tmp.key.hashCode() % newTable.length;
+                    if (index < 0) {
+                        index = index * (-1);
+                    }
+                    newTable[index] = tmp;
+                }
+
+                for (Node<K, V> current = tmp; current != null; current = current.next) {
+                    index = current.key.hashCode() % newTable.length;
+                    if (index < 0) {
+                        index = index * (-1);
+                    }
+                    newTable[index] = tmp;
+
+                }
+            }
+        }
+
+        threshold = (int) (newTable.length * loadFactor);
+        return tables = newTable;
+    }
+
+    private int index(K key) {
+        int h = (key == null) ? 0 : (key.hashCode() % tables.length);
+        if (h < 0) {
+            return h * (-1);
+        }
+        return h;
+    }
+
+    static class Node<K, V> {
+        private final int hash;
```

Comment:
```suggestion
```
Why store the hash in a separate variable if it can always be calculated?

Reply:
this is for convenience